concert/directors/base.py

```diff
@@ -0,0 +1,151 @@
+            experiment execution. When set() the next iteration will be prepared while the
+            experiment is still running. This could be used to prepare a future iteration while
+            still data is stored or processed.
+            The separate_scans property of the experiment should be set to False, since the director
+             handles the naming of the sub-folders.
+        :type experiment: concert.experiments.base.Experiment
+        """
+        self._experiment = experiment
+        self._run_event = asyncio.Event()
+        # Let us run by default
+        self._run_event.set()
+        self._iteration = 0
+        self.log = LOG
+        self.log.setLevel("INFO")
+        super().__init__()
+
+    async def _get_log_level(self):
+        return logging.getLevelName(self.log.getEffectiveLevel()).lower()
+
+    async def _set_log_level(self, level):
+        self.log.setLevel(level.upper())
+
+    async def _prepare_run(self, iteration: int):
+        """
+        This function changes whatever should be different between the different experiment
+        executions.
+
+        This could trigger an exchange of the specimen or adjust parameters of the experiment.
+        :param iteration:
+        :type iteration: int
+        """
+        raise NotImplementedError
+
+    async def _get_number_of_iterations(self) -> int:
+        """
+        Should return the total number of iterations.
+        :return: Number of iterations.
+        """
+        raise NotImplementedError
+
+    async def _get_iteration_name(self, iteration: int) -> str:
+        """
+        Function for giving meaningfully names for each experiment execution.
+        Should be overwritten for more complicated naming (e.g. specimen descriptions).
+        :param iteration:
+        :type  iteration: int
+        :return:
+        """
+        return f"iteration_{iteration:04d}"
+
+    async def _prepare_next_run(self):
+        if await self.get_current_iteration() < await self.get_number_of_iterations() - 1:
+            await self._prepare_run(await self.get_current_iteration() + 1)
+
+    @background
+    @check(source=['standby', 'error'], target="standby")
+    @transition(immediate="running", target="standby")
+    async def run(self):
+        handler = None
+        try:
+            if self._experiment.walker:
+                handler = logging.FileHandler(os.path.join(self._experiment.walker.current,
+                                                           'director.log'))
+                formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s '
+                                              '- %(message)s')
+                handler.setFormatter(formatter)
+                self.log.addHandler(handler)
+            self.log.info(await self.info_table)
+
+            await self.prepare()
+            # prepare first iteration
+            self._iteration = 0
+            await self._prepare_run(self._iteration)
```

Comment:
or `self._iteration = -1` and use the `_prepare_next_run` like for the rest of the iteration?

Reply:
I like it more this way. I thinks it is better to understand than calling `_prepare_run(0 - 1 + 1)`.